File: ratchet_aead.py
"""
ratchet_aead.py — FORWARD-SECRET session AEAD (wires item A, the auto-rekey ratchet, into the shell).

THE PROBLEM THIS SOLVES
-----------------------
aead.py and siv.py seal every message under the same master key. That is fine until the key leaks —
and then EVERY message ever sent under it, past and future, is readable. Real conversations want
FORWARD SECRECY: if the key is stolen today, yesterday's messages stay secret.

ratchet.py already built the engine for this — a one-way HMAC key chain that re-keys and burns the
old key — but nothing in the shell used it. This module wires it in at MESSAGE granularity: a session
of many messages where each message gets its own key from the chain, and the key (and the chain link
that made it) is burned once used. Capture the live session state after message N and you can read
message N onward (that is the live state) but NOT messages 0..N-1 — their keys are gone and the chain
is one-way, so they cannot be recomputed.

HOW IT COMPOSES WHAT WE ALREADY HAVE
------------------------------------
  * The forward-secret CHAIN is the same construction as ratchet.py (HMAC-SHA256, one-way, burn the
    old link). Here it advances once PER MESSAGE instead of per 64 KiB.
  * Each message is sealed with the ordinary committing AEAD (aead.seal / #6), keyed by that message's
    chain key. So every message inherits confidentiality + integrity + key-commitment for free; this
    module only adds the forward-secret keying on top.

      chain_0          = HMAC(master_key, "…|init|" ‖ nonce)
      msg_key_i        = HMAC(chain_i,    "…|msg|"  ‖ i)      # this message's AEAD key
      chain_{i+1}      = HMAC(chain_i,    "…|chain|"‖ i)      # next link; then BURN chain_i
      wire_i           = i(8, big-endian) ‖ aead.seal(msg_key_i, plaintext, aad ‖ i)

The message index i is sealed INTO the inner AEAD's aad, so tampering with the wire index makes the
inner open fail. The receiver advances its own chain in lockstep, burning links as it goes.

ORDER / SYNC (honest scope): messages must be processed in order (0,1,2,…); a receiver that has
advanced past i has BURNED key i and cannot go back — that is the whole point of forward secrecy, not
a bug. Gaps are tolerated by fast-forwarding (and burning) the skipped links. Out-of-order or
best-effort transports need a buffering layer on top; out of scope for this research artifact.

BURNING KEYS — same honest limit as ratchet.py: Python `bytes` are immutable + GC'd, so burning is
best-effort (drop the reference); the Rust core zeroizes in place. The CONSTRUCTION is forward-secret;
the in-memory GUARANTEE comes with the port.

STILL UNVETTED — adds forward secrecy to the shell; not a proof the chaos keystream is secure.
Validated in attacks/ratchet_aead_attack.py.
"""
from __future__ import annotations

from aead import InvalidTag, open_, seal
from ratchet import _kdf

_V = b"chaos-ratchet-aead-v1|"


def _init_chain(master_key: bytes, nonce: bytes) -> bytes:
    """chain_0, derived from the master key so the raw secret never keys a message directly."""
    return _kdf(bytes(master_key), _V + b"init|" + bytes(nonce))


def _derive(chain: bytes, index: int) -> tuple[bytes, bytes]:
    """From chain_i return (msg_key_i, chain_{i+1}). The caller burns chain_i afterwards."""
    idx = index.to_bytes(8, "big")
    return _kdf(chain, _V + b"msg|" + idx), _kdf(chain, _V + b"chain|" + idx)


def _bind_aad(aad: bytes, index: int) -> bytes:
    """Authenticate the message index inside the inner AEAD: length-prefix the caller's aad, then
    append the index. A tampered wire index won't match here -> inner open fails."""
    return len(aad).to_bytes(8, "big") + aad + index.to_bytes(8, "big")


class _Session:
    """Shared chain state for a forward-secret session. Sender and receiver each hold one and advance
    it in lockstep."""

    def __init__(self, master_key: bytes, nonce: bytes, aad: bytes = b""):
        if not isinstance(master_key, (bytes, bytearray)):
            raise TypeError("master_key must be bytes")
        self._aad = bytes(aad)
        self._chain = _init_chain(master_key, nonce)
        self._index = 0

    @property
    def index(self) -> int:
        """The next message index this session will produce / expect."""
        return self._index

    def _step(self) -> tuple[int, bytes]:
        """Advance one link: return (this index, its message key) and burn the consumed chain link."""
        i = self._index
        msg_key, next_chain = _derive(self._chain, i)
        self._chain = next_chain          # burn chain_i: overwrite the only reference (best-effort)
        self._index += 1
        return i, msg_key


class SenderSession(_Session):
    """Seal a sequence of messages with per-message forward secrecy."""

    def seal(self, plaintext: bytes, inner_nonce: bytes | None = None) -> bytes:
        """Seal the next message. Returns index(8) || committing-AEAD blob. After this returns, the
        key for THIS message is gone from the session — only the live message onward is recoverable.

        `inner_nonce` is optional and only for determinism in a known-answer / Rust-parity vector
        (parallels aead.seal's `nonce=`). In production leave it None: the inner AEAD draws a fresh
        random nonce. It is safe to pin here because every message already gets a UNIQUE key from the
        one-way chain, so a fixed inner nonce can never cause keystream reuse across messages."""
        i, msg_key = self._step()
        blob = seal(msg_key, plaintext, aad=_bind_aad(self._aad, i), nonce=inner_nonce)
        del msg_key
        return i.to_bytes(8, "big") + blob


class ReceiverSession(_Session):
    """Open a sequence of messages sealed by a SenderSession, advancing the chain in lockstep."""

    def open(self, wire: bytes) -> bytes:
        """Open the next message. Raises InvalidTag on tampering / wrong key, or ValueError if the
        index is in the past (its key was already burned — forward secrecy working as intended)."""
        if len(wire) < 8:
            raise InvalidTag("session message too short / malformed")
        i = int.from_bytes(wire[:8], "big")
        if i < self._index:
            raise ValueError(f"message {i} is in the past — its key was burned (forward secrecy)")
        # Tolerate gaps: fast-forward (and burn) any skipped links so we land on chain_i.
        while self._index < i:
            self._step()
        j, msg_key = self._step()
        assert j == i
        try:
            return open_(msg_key, wire[8:], aad=_bind_aad(self._aad, i))
        finally:
            del msg_key


if __name__ == "__main__":
    master = b"forward-secret session master key"
    nonce = b"session-nonce-001"
    convo = [b"hello", b"the package is in locker 12", b"code is 4471", b"burn after reading"]

    alice = SenderSession(master, nonce)
    bob = ReceiverSession(master, nonce)
    wires = [alice.seal(m) for m in convo]
    opened = [bob.open(w) for w in wires]
    print(f"session round-trip: {opened == convo}")

    # Forward secrecy: a fresh receiver that has advanced past message 1 cannot reopen message 0.
    late = ReceiverSession(master, nonce)
    late.open(wires[0]); late.open(wires[1])          # now poised at index 2; keys 0,1 burned
    try:
        late.open(wires[0])
        print("PAST MESSAGE RE-OPENED  <-- forward secrecy broken")
    except ValueError as e:
        print(f"past message protected: {e}")

File: rust/src/ratchet_aead.rs
//! Ratchet session AEAD — forward-secret SESSION over the committing AEAD (Phase 8.3).
//!
//! Port of ratchet_aead.py. A session where each message gets its OWN key from
//! a one-way HMAC-SHA256 chain; the link that made a key is burned after use.
//! Each message is sealed with the ordinary committing AEAD, inheriting
//! confidentiality + integrity + key-commitment for free.
//!
//! Wire format: index(8,BE) || aead_seal(msg_key_i, inner_nonce_i, plaintext, bind_aad)
//!
//! STILL UNVETTED. See REPORT.md.

use zeroize::{Zeroize, Zeroizing};

use crate::aead::{aead_open, aead_seal};
use crate::utils::{cat, hmac_sha256};

/// Domain-separation tag for the session-AEAD chain.
const RATCHET_AEAD_V: &[u8] = b"chaos-ratchet-aead-v1|";

/// chain_0, derived from the master key so the raw secret never keys a message directly.
fn ra_init_chain(master_key: &[u8], nonce: &[u8]) -> [u8; 32] {
    hmac_sha256(master_key, &cat(&[RATCHET_AEAD_V, b"init|", nonce]))
}

/// From chain_i return (msg_key_i, chain_{i+1}). The caller burns chain_i afterwards.
fn ra_derive(chain: &[u8; 32], index: u64) -> ([u8; 32], [u8; 32]) {
    let idx = index.to_be_bytes();
    let msg_key = hmac_sha256(chain, &cat(&[RATCHET_AEAD_V, b"msg|", &idx]));
    let next_chain = hmac_sha256(chain, &cat(&[RATCHET_AEAD_V, b"chain|", &idx]));
    (msg_key, next_chain)
}

/// Authenticate the message index inside the inner AEAD: length-prefix the aad, then append the index.
fn ra_bind_aad(aad: &[u8], index: u64) -> Vec<u8> {
    let alen = (aad.len() as u64).to_be_bytes();
    cat(&[&alen, aad, &index.to_be_bytes()])
}

/// Shared chain state for a forward-secret session. Sender and receiver each hold one.
struct RatchetAeadSession {
    aad: Vec<u8>,
    chain: Zeroizing<[u8; 32]>,
    index: u64,
    n_maps: usize,
}

impl RatchetAeadSession {
    fn new(master_key: &[u8], nonce: &[u8], aad: &[u8], n_maps: usize) -> Self {
        RatchetAeadSession {
            aad: aad.to_vec(),
            chain: Zeroizing::new(ra_init_chain(master_key, nonce)),
            index: 0,
            n_maps,
        }
    }

    /// Advance one link: return (this index, its message key) and BURN the consumed chain link.
    fn step(&mut self) -> (u64, Zeroizing<[u8; 32]>) {
        let i = self.index;
        let (msg_key, mut next_chain) = ra_derive(&self.chain, i);
        self.chain.zeroize();
        self.chain.copy_from_slice(&next_chain);
        next_chain.zeroize();
        self.index += 1;
        (i, Zeroizing::new(msg_key))
    }
}

/// Seal a sequence of messages with per-message forward secrecy.
pub struct RatchetAeadSender(RatchetAeadSession);

impl RatchetAeadSender {
    pub fn new(master_key: &[u8], nonce: &[u8], aad: &[u8], n_maps: usize) -> Self {
        RatchetAeadSender(RatchetAeadSession::new(master_key, nonce, aad, n_maps))
    }

    /// Seal the next message. Returns index(8,BE) || committing-AEAD blob.
    pub fn seal(&mut self, inner_nonce: &[u8], plaintext: &[u8]) -> Vec<u8> {
        let (i, msg_key) = self.0.step();
        let aad = ra_bind_aad(&self.0.aad, i);
        let blob = aead_seal(&msg_key[..], inner_nonce, plaintext, &aad, self.0.n_maps);
        let mut wire = Vec::with_capacity(8 + blob.len());
        wire.extend_from_slice(&i.to_be_bytes());
        wire.extend_from_slice(&blob);
        wire
    }
}

/// Open a sequence of messages sealed by a `RatchetAeadSender`, advancing in lockstep.
pub struct RatchetAeadReceiver(RatchetAeadSession);

impl RatchetAeadReceiver {
    pub fn new(master_key: &[u8], nonce: &[u8], aad: &[u8], n_maps: usize) -> Self {
        RatchetAeadReceiver(RatchetAeadSession::new(master_key, nonce, aad, n_maps))
    }

    /// Open the next message. None on tamper / wrong key, or if the index is in the past.
    pub fn open(&mut self, wire: &[u8]) -> Option<Vec<u8>> {
        if wire.len() < 8 {
            return None;
        }
        let i = u64::from_be_bytes(wire[..8].try_into().ok()?);
        if i < self.0.index {
            return None; // key burned — forward secrecy
        }
        while self.0.index < i {
            let _ = self.0.step(); // fast-forward + burn skipped links
        }
        let (j, msg_key) = self.0.step();
        debug_assert_eq!(j, i);
        let aad = ra_bind_aad(&self.0.aad, i);
        aead_open(&msg_key[..], &wire[8..], &aad, self.0.n_maps)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::constants::DEFAULT_N_MAPS;

    #[test]
    fn ratchet_aead_session_roundtrip_and_forward_secrecy() {
        let master = b"a forward-secret session master key";
        let nonce = b"session-nonce-001";
        let aad = b"channel-7";
        let convo: [&[u8]; 3] = [b"hello", b"the package is in locker 12", b"burn after reading"];
        let inonces: [&[u8]; 3] = [b"inner-nonce-aaa0", b"inner-nonce-aaa1", b"inner-nonce-aaa2"];

        let mut tx = RatchetAeadSender::new(master, nonce, aad, DEFAULT_N_MAPS);
        let wires: Vec<Vec<u8>> = convo
            .iter()
            .zip(inonces.iter())
            .map(|(m, n)| tx.seal(n, m))
            .collect();

        let mut rx = RatchetAeadReceiver::new(master, nonce, aad, DEFAULT_N_MAPS);
        for (w, m) in wires.iter().zip(convo.iter()) {
            assert_eq!(rx.open(w).as_deref(), Some(*m));
        }

        // forward secrecy: a receiver past message 0/1 cannot reopen them
        let mut late = RatchetAeadReceiver::new(master, nonce, aad, DEFAULT_N_MAPS);
        assert!(late.open(&wires[0]).is_some());
        assert!(late.open(&wires[1]).is_some());
        assert!(late.open(&wires[0]).is_none(), "past message reopened — forward secrecy broken");
        assert!(late.open(&wires[1]).is_none());

        // gap tolerated: a fresh receiver jumps to message 2, but can't go back
        let mut skip = RatchetAeadReceiver::new(master, nonce, aad, DEFAULT_N_MAPS);
        assert_eq!(skip.open(&wires[2]).as_deref(), Some(convo[2]));
        assert!(skip.open(&wires[0]).is_none());
    }

    #[test]
    fn ratchet_aead_rejects_tamper_wrong_key_and_index() {
        let master = b"session master";
        let nonce = b"sess-nonce";
        let aad = b"";
        let inonce = b"sixteen-byte-non";
        let pt = b"secret session payload";

        let mut tx = RatchetAeadSender::new(master, nonce, aad, DEFAULT_N_MAPS);
        let wire = tx.seal(inonce, pt);

        assert_eq!(
            RatchetAeadReceiver::new(master, nonce, aad, DEFAULT_N_MAPS)
                .open(&wire)
                .as_deref(),
            Some(&pt[..])
        );
        // tamper an inner ciphertext byte
        let mut bad = wire.clone();
        bad[8 + crate::aead::NONCE_LEN + crate::aead::COMMIT_LEN + 1] ^= 0x01;
        assert!(RatchetAeadReceiver::new(master, nonce, aad, DEFAULT_N_MAPS)
            .open(&bad)
            .is_none());
        // wrong master key
        assert!(RatchetAeadReceiver::new(b"wrong", nonce, aad, DEFAULT_N_MAPS)
            .open(&wire)
            .is_none());
        // bump the wire index -> inner aad mismatch -> open fails
        let mut idx = wire.clone();
        idx[7] ^= 0x01;
        assert!(RatchetAeadReceiver::new(master, nonce, aad, DEFAULT_N_MAPS)
            .open(&idx)
            .is_none());
    }
}
